src/main/scala-2.11/markets/clearing/engines/CDAMatchingEngine.scala

```diff
@@ -15,24 +15,208 @@ limitations under the License.
+          matches.enqueue(totalMatch)
+        } else {  // incoming order is larger than existing order and will be rationed!
+          val partialMatch = PartialMatch(askOrder, incoming, price)
+          val residualOrder = incoming.split(residualQuantity)
+          findMatchingAskOrders(residualOrder, matches.enqueue(partialMatch))
+        }
+
+      case _ => // existingOrders is empty or incoming order does not cross best existing order.
+        _bidOrderBook.add(incoming)  // SIDE EFFECT!
+        matches
+    }
+  }
+
+  @tailrec
+  private def findMatchingBidOrders(incoming: AskOrder,
+                                    matches: immutable.Queue[Match]): immutable.Queue[Match] = {
+    _bidOrderBook.headOption match {
+      case Some(bidOrder) if crosses(incoming, bidOrder) =>
+
+        _bidOrderBook.remove(bidOrder)  // SIDE EFFECT!
+        val residualQuantity = incoming.quantity - bidOrder.quantity
+        val price = formPrice(incoming, bidOrder)
+
+        if (residualQuantity < 0) {
+          val totalMatch = TotalMatch(bidOrder, incoming, price)
+          // add residualOrder back into orderBook!
+          val residualOrder = bidOrder.split(-residualQuantity)
+          _bidOrderBook.add(residualOrder)  // SIDE EFFECT!
+          matches.enqueue(totalMatch)
+        } else if (residualQuantity == 0) {  // no rationing for incoming order!
+          val totalMatch = TotalMatch(bidOrder, incoming, price)
+          matches.enqueue(totalMatch)
+        } else {  // incoming order is larger than existing order and will be rationed!
+          val partialMatch = PartialMatch(bidOrder, incoming, price)
+          val residualOrder = incoming.split(residualQuantity)
+          findMatchingBidOrders(residualOrder, matches.enqueue(partialMatch))
+        }
+      case _ => // existingOrders is empty or incoming order does not cross best existing order.
+        _askOrderBook.add(incoming)  // SIDE EFFECT!
+        matches
+    }
+  }
+
+  /** Implements price formation rules for limit and market orders.
+    *
+    * This matching engine uses the “Best limit” price improvement rule: if the opposite book
+    * does have limit orders, then the trade settles at the better of two prices (either the
+    * incoming order’s limit or the best limit from the opposite book) the term “better of two
+    * prices” is from the point of view of the incoming limit order. In other words, if incoming
+    * limit order would have crossed with outstanding opposite “best limit” order in the absence
+    * of market order, then the trade would execute at that, potentially improved, “best limit”
+    * price.
+    *
+    * @param incoming the incoming order.
+    * @param existing the order that resides at the top of the opposite book.
+    * @return the price at which the trade between the two orders will execute.
+    */
+  def formPrice(incoming: Order, existing: Order): Long = {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Cyclomatic complexity of 15 exceeds max of 10](https://www.codacy.com/app/davidrpugh/markets-sandbox/file/1519375750/issues/source?bid=2921361&fileBranchId=3008855#l151)